doc/developer/platform/ux.md

```diff
@@ -772,6 +772,15 @@ running will be hidden from users. Any log messages in `materialized` today that
 are meant for end-users will need to instead make their messages available via
 tables or sources in the system catalog.
 
+### Correctness
+
+Materialize aims to maintain three correctness guarantees.
+
+1. *Materialize is serializable (or strictly serializable) against all internal reads and writes.*
```

Comment:
Are there cases where we are _serializable_ and not _strict serializable_? Unless I am interpreting "all internal reads and writes" wrong (I am interpreting it as: "reads and writes against views involving only tables and sources that have not changed", and we should define it here), then I think the answer is no?

Reply:
At the moment we will happily answer queries at timestamps that go backwards. We would need to stop doing that if we wanted strict serializability. But I agree, a good moment to decide that (in that we should think hard, rather than flip a coin).

Reply:
That choice feels consequential.  Should we have it here for posterity?

Reply:
Not sure I follow! Should we have a discussion of the choice in the doc for posterity?

I like the idea of supporting *both* serializability and strict serializability as modes, and punting on the decision for what the default until we have a better understanding of the performance characteristics of both modes.

Reply:
Recording live discussion for posterity: we want to try to suss out what "internal read and write" means here!

Reply:
I think we were moving towards "reads and writes against Materialize tables" but I might have missed a detail...

Reply:
Maybe:

> Transactions in Materialize are strictly serializable with respect to the operations that occur inside of Materialize. Operations include:
  * `SELECT`, `INSERT`, UPDATE`, and `DELETE` statements (but not `TAIL`)
  * The commit of an offset by a Kafka source

Reply:
To be extra specific, I think @benesch's language makes the stronger claim that even reads from external sources are included in the set of things that are strictly serializable, which is "fine" because we do not include writes to those sources in that set of things. We are not claiming to form a strictly serializable system with the upstream sources, which would be hard, but that all of your interactions with Materialize can be strictly serialized.

Reply:
@benesch can you say more about the second bullet -- "The commit of an offset by a Kafka source" -- that's essentially bringing strict serializability to at least one transaction outside of Materialize and it feels a little asymmetric -- I also worry that there might be some future encoding-delivered-over-kafka that is more gnarly than debezium that we just won't be able to handle -- so we might want to be very specific about which Kafka sources here?

Reply:
> @benesch can you say more about the second bullet -- "The commit of an offset by a Kafka source" -- that's essentially bringing strict serializability to at least one transaction outside of Materialize and it feels a little asymmetric -- I also worry that there might be some future encoding-delivered-over-kafka that is more gnarly than debezium that we just won't be able to handle -- so we might want to be very specific about which Kafka sources here?

So committing an offset to Kafka is the way that a consumer (i.e., Materialize) indicates that it has successfully read from Kafka. PostgreSQL actually has something similar; I've updated my comment above accordingly. The way to think about it is that several Materialize sources can communicate _back upstream_ that Materialize has successfully ingested data up to some offset. Materialize is strictly serializable wrt these acknowledgements, too.

These sorts of acknowledgements should be pretty unfazed by what encoding is being transported over the source. If we can reclock it, we can send these acks upstream. Basically we want to commit these Kafka/Postgres offsets upstream as soon as we have durably recorded timestamp bindings.

Reply:
Ah!  I understand -- so Materialize initiated.  An ACK!  

Is this something that is a bit more general to sources?  Like

* Materialize initiated acknowledgements of upstream sources (e.g., the commit of an offset by a Kafka source, the acknowledge of an LSN by a PostgresSQL source, etceterea)

Reply:
Yeah, that's exactly right!